provisioner/terraform/timings.go

```diff
@@ -0,0 +1,238 @@
+// the total time taken to perform a certain action.
+// The resulting slice of *proto.Timing is NOT sorted.
+func (t *timingAggregator) aggregate() []*proto.Timing {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	out := make([]*proto.Timing, 0, len(t.stateLookup))
+
+	for _, s := range t.stateLookup {
+		// We are only concerned here with failed or completed events.
+		if s.state != proto.TimingState_FAILED && s.state != proto.TimingState_COMPLETED {
+			continue
+		}
+
+		// Look for a corresponding span for the STARTED state.
+		startSpan, ok := t.stateLookup[s.hashByState(proto.TimingState_STARTED)]
+		if !ok {
+			// Not found, we'll ignore this span.
+			continue
+		}
+		s.start = startSpan.start
+
+		// Until faster-than-light travel is a possibility, let's prevent this.
+		// Better to capture a zero delta than a negative one.
+		if s.start.After(s.end) {
+			s.start = s.end
+		}
+
+		// Let's only aggregate valid entries.
+		// Later we can add support for partial / failed applies, perhaps.
+		if s.start.IsZero() || s.end.IsZero() {
+			continue
+		}
+
+		out = append(out, s.toProto())
+	}
+
+	return out
+}
+
+func (l timingKind) Valid() bool {
+	return slices.Contains([]timingKind{
+		timingApplyStart,
+		timingApplyProgress,
+		timingApplyComplete,
+		timingApplyErrored,
+		timingProvisionStart,
+		timingProvisionProgress,
+		timingProvisionComplete,
+		timingProvisionErrored,
+		timingRefreshStart,
+		timingRefreshComplete,
+		timingChangeSummary,
+		timingDiagnostic,
+		timingPlannedChange,
+		timingOutputs,
+		timingResourceDrift,
+		timingVersion,
+		timingInitStart,
+		timingInitComplete,
+		timingInitErrored,
+		timingGraphStart,
+		timingGraphComplete,
+		timingGraphErrored,
+		timingLog,
+		timingInitOutput,
+	}, l)
+}
+
+// Category returns the category for a giving timing state so that timings can be aggregated by this category.
+// We can't use the state itself because we need an `apply_start` and an `apply_complete` to both hash to the same entry
+// if all other attributes are identical.
+func (l timingKind) Category() string {
+	switch l {
+	case timingInitStart, timingInitComplete, timingInitErrored:
+		return "init"
+	case timingGraphStart, timingGraphComplete, timingGraphErrored:
+		return "graph"
+	case timingApplyStart, timingApplyProgress, timingApplyComplete, timingApplyErrored:
+		return "apply"
+	case timingProvisionStart, timingProvisionProgress, timingProvisionComplete, timingProvisionErrored:
+		return "provision"
+	case timingRefreshStart, timingRefreshComplete:
+		return "state refresh"
+	default:
+		return "?"
+	}
+}
+
+// hashState computes a hash based on a timingSpan's unique properties and state.
+// The combination of resource and provider names MUST be unique across entries.
+func (e *timingSpan) hashByState(state proto.TimingState) uint64 {
+	id := fmt.Sprintf("%s:%s:%s:%s", e.kind.Category(), state.String(), e.resource, e.provider)
```

Comment:
nit: strings.Builder?

Reply:
Agreed but it's less concise and this is clear enough IMHO.